allocator/nonmoving_unix.go

```diff
@@ -41,46 +40,39 @@ func alloc(_, max uint64) experimental.LinearMemory {
 type mmappedMemory struct {
 	buf []byte
 	max uint64
-
-	// Any reasonable Wasm implementation will take a lock before calling Grow, but this
-	// is invisible to Go's race detector so it can still detect raciness when we updated
-	// buf. We go ahead and take a lock when mutating since the performance effect should
-	// be negligible in practice and it will help the race detector confirm the safety.
-	mu sync.Mutex
 }
 
 func (m *mmappedMemory) Reallocate(size uint64) []byte {
 	if size > m.max {
-		panic(errInvalidReallocation)
+		return nil
 	}
 
-	m.mu.Lock()
-	defer m.mu.Unlock()
-
 	com := uint64(len(m.buf))
-	if com < size {
+	res := uint64(cap(m.buf))
+
+	if com < size && size <= res {
 		// Round up to the page size.
-		rnd := uint64(unix.Getpagesize() - 1)
+		rnd := pageSize - 1
 		newCap := (size + rnd) &^ rnd
 
 		// Commit additional memory up to new bytes.
 		err := unix.Mprotect(m.buf[com:newCap], unix.PROT_READ|unix.PROT_WRITE)
 		if err != nil {
-			panic(fmt.Errorf("allocator_unix: failed to commit memory: %w", err))
+			return nil
 		}
 
 		// Update committed memory.
 		m.buf = m.buf[:newCap]
 	}
 	// Limit returned capacity because bytes beyond
 	// len(m.buf) have not yet been committed.
+	if size < uint64(len(m.buf)) {
```

Comment:
@ncruces This is notable new logic. In https://github.com/wasilibs/go-re2/issues/188 I notice size is greater than the length of the buffer - that being said I'm bad at math, could there be an issue with `newCap` computation above?